File: app/src/main/java/dev/mike/couchtour/CatalogCache.kt
package dev.mike.couchtour

/**
 * How long a cached catalog entry (periods, shows, show detail) stays fresh before the next
 * read re-fetches it (#61). Long enough that ordinary back-navigation within a session costs
 * no network round trip; short enough that a show's rating, popularity, or tags don't go
 * stale for the length of a whole sitting.
 */
internal const val CATALOG_CACHE_TTL_MS = 15 * 60 * 1000L

/**
 * A small in-memory, TTL-bounded cache shared by both [MusicSource] implementations (#61).
 * In-memory rather than on-disk was the explicit call for this pass — it's reversible and
 * needs no Room migration, unlike a persisted catalog cache would (see DECISIONS.md).
 *
 * [maxEntries] is the memory ceiling: a ~200-artist catalog with years and shows underneath
 * it means an unbounded shows/show-detail cache could grow without limit as a session goes
 * on, so the map evicts its least-recently-used entry once full rather than growing forever.
 *
 * Not private to its callers' fields, the same reasoning as
 * [RelistenCatalogSource.cachedArtists]: tests reset state between runs by calling [clear].
 */
internal class TtlCache<K, V>(
    private val ttlMillis: Long,
    private val maxEntries: Int,
    private val clock: () -> Long = System::currentTimeMillis,
) {
    private data class Entry<V>(val value: V, val storedAt: Long)

    // LinkedHashMap in access order gives LRU eviction for free; synchronized because
    // MusicSource calls run from concurrent coroutines (e.g. NextStop fetching several
    // periods' shows in parallel).
    private val entries = object : LinkedHashMap<K, Entry<V>>(16, 0.75f, true) {
        override fun removeEldestEntry(eldest: MutableMap.MutableEntry<K, Entry<V>>) =
            size > maxEntries
    }

    @Synchronized
    fun get(key: K): V? {
        val entry = entries[key] ?: return null
        if (clock() - entry.storedAt >= ttlMillis) {
            entries.remove(key)
            return null
        }
        return entry.value
    }

    @Synchronized
    fun put(key: K, value: V) {
        entries[key] = Entry(value, clock())
    }

    @Synchronized
    fun clear() {
        entries.clear()
    }

    @get:Synchronized
    val size: Int
        get() = entries.size
}

File: macos/Packages/CouchTourKit/Sources/CouchTourKit/CatalogCache.swift
import Foundation

/// How long a cached catalog entry (periods, shows, show detail) stays fresh before the next
/// read re-fetches it (#61). Long enough that ordinary back-navigation within a session costs
/// no network round trip; short enough that a show's rating, popularity, or tags don't go
/// stale for the length of a whole sitting. Port of `CatalogCache.kt`'s
/// `CATALOG_CACHE_TTL_MS`.
let catalogCacheTTL: TimeInterval = 15 * 60

/// A small in-memory, TTL-bounded cache shared by both `MusicSource` implementations (#61).
/// In-memory rather than on-disk was the explicit call for this pass — it's reversible and
/// needs no GRDB migration, unlike a persisted catalog cache would (see DECISIONS.md).
///
/// `maxEntries` is the memory ceiling: a ~200-artist catalog with years and shows underneath
/// it means an unbounded shows/show-detail cache could grow without limit as a session goes
/// on, so the store evicts its least-recently-used entry once full rather than growing
/// forever. An `actor` because, unlike `RelistenCatalogSource` (itself already an actor),
/// `PhishInSource` is a plain struct recreated on every `sourceFor` call — its cache has to
/// live somewhere shared and thread-safe on its own.
actor TTLCache<Key: Hashable & Sendable, Value: Sendable> {
    private struct Entry {
        let value: Value
        let storedAt: Date
    }

    private var entries: [Key: Entry] = [:]
    /// Access order, oldest first, for LRU eviction — the same shape as Android's
    /// `LinkedHashMap(accessOrder = true)`.
    private var order: [Key] = []
    private let ttl: TimeInterval
    private let maxEntries: Int
    private let now: () -> Date

    init(ttl: TimeInterval, maxEntries: Int, now: @escaping () -> Date = Date.init) {
        self.ttl = ttl
        self.maxEntries = maxEntries
        self.now = now
    }

    func get(_ key: Key) -> Value? {
        guard let entry = entries[key] else { return nil }
        if now().timeIntervalSince(entry.storedAt) >= ttl {
            entries.removeValue(forKey: key)
            order.removeAll { $0 == key }
            return nil
        }
        order.removeAll { $0 == key }
        order.append(key)
        return entry.value
    }

    func put(_ key: Key, _ value: Value) {
        if entries[key] == nil {
            order.append(key)
        } else {
            order.removeAll { $0 == key }
            order.append(key)
        }
        entries[key] = Entry(value: value, storedAt: now())
        while order.count > maxEntries {
            let evicted = order.removeFirst()
            entries.removeValue(forKey: evicted)
        }
    }

    func clear() {
        entries.removeAll()
        order.removeAll()
    }

    var count: Int { entries.count }
}
